text/0026-block-version-based-protocol-evolution.md

```diff
@@ -0,0 +1,288 @@
+| ------------- | ---------                 | -----------                                                                                                                                                  |
+| 1             |    -                      | Status quo                                                                                                                                                   |
+| 2             | `TxOut::e_memo`           | This is the first block version at which `e_memo` field will be accepted, and it will be required going forwards. This represents the changes in [MCIP #3](https://github.com/mobilecoinfoundation/mcips/pull/0003).    |
+| 3             | `Amount::masked_token_id` | This is the first block version at which `masked_token_id` will be accepted, and it will be required going forward. This represents the changes in [MCIP #25](https://github.com/mobilecoinfoundation/mcips/pull/0025). |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+
+We'll attempt to describe code changes that would be needed in mobilecoin software to implement this:
+
+The current block version can be obtained from the `mc-ledger-db` LMDB ledger interface by getting
+the block header of the latest block, so clients like `mobilecoind` and `full-service` can already access this number.
+
+The changes that are required are:
+* Fog should start telling mobile clients about the latest block version, similarly to how it currently advertises the
+"highest known block count". Both fog-view and fog-ledger should do this.
+* The standard transaction builder should become `block_version` aware. Clients should pass the latest block version to it.
+* Transaction validation should become `block_version` aware.
+* The consensus enclave should support dynamic `block_version`.
+
+Mobile clients SHOULD use the `latest_block_version` number to indicate to the users that an upgrade is required. The constant
+`mc_transaction_core::MAX_BLOCK_VERSION` will indicate the largest block version that this version of the `TransactionBuilder` object supports.
+
+Desktop clients, which use the lmdb database `mc-ledger-db`, already work by comparing the block version of new blocks to the
+`mc_transaction_core::BLOCK_VERSION` constant. They should continue doing this, but the constant should be renamed `MAX_BLOCK_VERSION`.
+
+We do not at this time propose that the `Tx` object itself should contain a `block_version`.
+Instead, transactions are valid if they conform to the rules for the current block version.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This adds complexity to the transaction validation code, which now must support multiple versions of the rules.
```

Comment:
idk, I guess my thinking is:
* In the "initial" plans for how to do versioning, we didn't plan on doing this, there would just be one version for transaction core lib.
* We also planned to roll out the memos that way
* I'm arguing that we have to accept this complexity to avoid breaking the clients sometimes in this MCIP but I'm trying to treat that as a position of the MCIP rather than state it as a fact, and the reader can draw their conclusions